### PR TITLE
scim2 guides

### PR DESCRIPTION
**Pull request by nilminiwso2**
## Purpose
Adding examples on SCIM2.

### CODE REVIEW

File: en/docs/develop/scim2-patch-operations.md [hunk 1]
```diff
@@ -0,0 +1,928 @@
...
+        <td>A resource attribute that contains a single value. For example, the <code>displayName</code> attribute can only have one unique value.
+        </td>
+    </tr>
+    <tr>
+        <th>Multi-valued attribute</th>
+        <td>A resource attribute that contains multiple values. For example, the <code>emails</code> attribute may have multiple email addresses as values.
+        </td>
+    </tr>
+    <tr>
+        <th>Complex attributes</th>
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:28]:
```suggestion
        <th>Complex attribute</th>
```

File: en/docs/develop/scim2-patch-operations.md [hunk 2]
```diff
@@ -0,0 +1,928 @@
...
+        <td>A singular or multi-valued attribute, which has a value that is a composition of one or more simple attributes. For example, the <code>addresses</code> attribute has the sub-attributes <code>streetAddress</code>, <code>locality</code>, <code>postalCode</code>, and <code>country</code>.
+        </td>
+    </tr>
+    <tr>
+        <th>Simple attributes</th>
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:33]:
```suggestion
        <th>Simple attribute</th>
```

File: en/docs/develop/scim2-patch-operations.md [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+        <td>Required if <code>op</code> is <code>remove</code>.</br>
+        Optional if <code>op</code> is <code>add</code> or <code>replace</code>.
+        </td>
+        <td>Add the path to specify the attribute/sub-attribute that should be updated.</br> 
+        Since <code>path</code> is an optional parameter for <code>add</code> and <code>replace</code> operations, there are two ways to define the operation in a payload: with or without the <code>path</code> parameter.
+        </td>
+    </tr>
+    <tr>
+        <td><code>value</code></td>
+        <td>Required</td>
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:127]: value is required only if op is "add" or "replace"

File: en/docs/develop/scim2-patch-operations.md [hunk 4]
```diff
@@ -0,0 +1,928 @@
...
+        "op": "add",
+        "value": {
+            "emails": [
+                {
+                    "value": "abc@gmail.com",
+                    "type": "home"
+                },
+                {
+                "value": "xyz@gmail.com",
+                "type": "work"
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:313]: Formatting issue. Need an indentation

File: en/docs/develop/scim2-patch-operations.md [hunk 5]
```diff
@@ -0,0 +1,928 @@
...
+
+Let's consider a custom schema with an attribute called devices.
+
+-   **Example 1:** Replace the `devices` attribue in a **custom** user schema (without using the path param):
+
+    ```json
+    {
+        "op": "replace",
+        "value": {
+            "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User": {
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:471]:
```suggestion
            "urn:scim:wso2:schema": {
```

File: en/docs/develop/scim2-patch-operations.md [hunk 6]
```diff
@@ -0,0 +1,928 @@
...
+        }
+    }
+    ```
+
+-   **Example 2:** Replace the `devices` attribue in a **custom** user schema (using the path param):
+
+    ```json
+    {
+        "op": "replace",
+        "path": "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:devices",
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:483]:
```suggestion
        "path": "urn:scim:wso2:schema:devices",
```

File: en/docs/develop/scim2-patch-operations.md [hunk 7]
```diff
@@ -0,0 +1,928 @@
...
+            "M7"
+        ]
+    }
+    ```
+
+**Complex multi-valued attributes**
+
+Consider the `emails` attribute in the **Core** user schema, which can have multiple emails types such as `home`, `work`, etc. Each email attribute has sub-attributes such as `value`, `type`, and `primary`.
+
+-   **Example 1:** Replace the `emails` attribute in a **custom** user schema (without using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:495]:
```suggestion
-   **Example 1:** Replace the `emails` attribute in a **Core** user schema (without using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 8]
```diff
@@ -0,0 +1,928 @@
...
+                {
+                    "value": "wxyz@gmail.com",
+                    "type": "work"
+                }
+            ]
+        }
+    }
+    ```
+
+-   **Example 2:** Replace the `emails` attribute in a **custom** user schema (using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:515]:
```suggestion
-   **Example 2:** Replace the `emails` attribute in a **Core** user schema (using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 9]
```diff
@@ -0,0 +1,928 @@
...
+    }
+    ```
+
+**Simple multi-valued attributes**
+
+This type of attribute is not found in the core schemas of the SCIM specification. However, there is a capability to add simple multi-valued attributes to our extended schemas (custom schemas). 
+
+Let's consider a custom schema with an attribute called devices.
+
+-   **Example 1:** Replace the `devices` attribue in a **custom** user schema (without using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:465]:
```suggestion
-   **Example 1:** Replace the `devices` attribue in a **Custom** user schema (without using the path param):
```
Let's capitalize the `C` to match with Core and Enterprise keywords. Check other places as well

File: en/docs/develop/scim2-patch-operations.md [hunk 10]
```diff
@@ -0,0 +1,928 @@
...
+    }
+    ```
+
+**Complex singular attributes**
+
+Consider the `name` attribute in **Core** user schema, which has sub-attributes such as `givenName`, `familyName`, etc. and the `manager` attribute in the **Enterprise** user schema (`urn:ietf:params:scim:schemas:extension:enterprise:2.0:User`), which has `displayName`, `value`, `ref`, etc. as sub-attributes. 
+
+!!! Note    
+    -   If the path contains the complex attribute, all sub-attribute values are removed. 
+    -   There is a special case for the `name` attribute. Even though `“path”: “name”` is specified, the `familyName` attribute is not removed.
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:566]: There is a special case for the `name` attribute. Even though `“path”: “name”` is specified, the `familyName` attribute is not removed. This special case is only for LDAP userstore. Let's mention that as well

File: en/docs/develop/scim2-patch-operations.md [hunk 11]
```diff
@@ -0,0 +1,928 @@
...
+-   **Example 1:** Remove a sub-attribute of the `name` attribute in the **Core** user schema:
+
+    ```json
+    {
+        "op": "remove",
+        "path": "name.givenName"
+    }
+    ```
+
+-   **Example 2:** Remove the `name` attribute in the **Enterprise** user schema:
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:578]:
```suggestion
-   **Example 2:** Remove the `manager's value` attribute and `whole manager` object in the **Enterprise** user schema:
```

File: en/docs/develop/scim2-patch-operations.md [hunk 12]
```diff
@@ -0,0 +1,928 @@
...
+**Simple multi-valued attributes**
+
+This type of attribute is not found in the core schemas of the SCIM specification. However, there is a capability to add simple multi-valued attributes to our extended schemas (custom schemas). 
+
+Let's consider a custom schema with an attribute called devices.
+
+```json
+{
+    "op": "remove",
+    "path": "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:devices[value eq \"M7\"]"
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:600]:
```suggestion
    "path": "urn:scim:wso2:schema:devices[value eq \"M7\"]"
```

File: en/docs/develop/scim2-patch-operations.md [hunk 13]
```diff
@@ -0,0 +1,928 @@
...
+},
+{
+    "op": "remove",
+    "path": "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User:devices"
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:604]:
```suggestion
    "path": "urn:scim:wso2:schema:devices"
```

File: en/docs/develop/scim2-patch-operations.md [hunk 14]
```diff
@@ -0,0 +1,928 @@
...
+    ```json
+    {
+        "op": "replace",
+        "path": "members",
+        "value": [
+                    {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:693]: formatting issue.

File: en/docs/develop/scim2-patch-operations.md [hunk 15]
```diff
@@ -0,0 +1,928 @@
...
+    {
+        "op": "add",
+        "path": "members",
+        "value": [
+                    {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
+    }
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:659]: formatting issue

File: en/docs/develop/scim2-patch-operations.md [hunk 16]
```diff
@@ -0,0 +1,928 @@
...
+    ```
+
+### Replace members
+
+Let's create patch request payloads to `replace` members in a user group.
+
+-   **Example 1:** Replace a member in a group (without using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:666]:
```suggestion
-   **Example 1:** Replace members in a group (without using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 17]
```diff
@@ -0,0 +1,928 @@
...
+                {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
+        }
+    }
+    ```
+
+-   **Example 2:** Replace a member in a group (using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:682]:
```suggestion
-   **Example 2:** Replace members in a group (using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 18]
```diff
@@ -0,0 +1,928 @@
...
+    ```json
+    {
+        "op": "add",
+        "path": "users",
+        "value": [
+                    {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:781]: formatting issue

File: en/docs/develop/scim2-patch-operations.md [hunk 19]
```diff
@@ -0,0 +1,928 @@
...
+    }
+    ```
+
+### Replace users
+
+Let's create patch request payloads to `replace` users assigned to a user role.
+
+-   **Example 1:** Replace a user in a role (without using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:789]:
```suggestion
-   **Example 1:** Replace users in a role (without using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 20]
```diff
@@ -0,0 +1,928 @@
...
+    ```json
+    {
+        "op": "replace",
+        "path": "users",
+        "value": [
+                    {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:816]: formatting issue

File: en/docs/develop/scim2-patch-operations.md [hunk 21]
```diff
@@ -0,0 +1,928 @@
...
+                {
+                    "display": "alex",
+                    "value": "0565f472-28fe-4d93-83ad-096c66ed4a47"
+                }
+            ]
+        }
+    }
+    ```
+
+-   **Example 2:** Replace a user in a role (using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:805]:
```suggestion
-   **Example 2:** Replace users in a role (using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 22]
```diff
@@ -0,0 +1,928 @@
...
+### Remove users
+
+Let's create patch request payloads to `remove` users from a user role.
+
+-   **Example 1:** Specify the user (by user ID) who needs to be removed.
+
+    ```json
+    {
+        "op": "remove",
+        "path": "members[value eq 0565f472-28fe-4d93-83ad-096c66ed4a47]"
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:829]:
```suggestion
        "path": "users[value eq 0565f472-28fe-4d93-83ad-096c66ed4a47]"
```

File: en/docs/develop/scim2-patch-operations.md [hunk 23]
```diff
@@ -0,0 +1,928 @@
...
+    }
+    ```
+
+-   **Example 2:** Specify the user (by username) who needs to be removed.
+
+    ```json
+    {
+        "op": "remove",
+        "path": "members[display eq alex]"
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:838]:
```suggestion
        "path": "members[users eq alex]"
```

File: en/docs/develop/scim2-patch-operations.md [hunk 24]
```diff
@@ -0,0 +1,928 @@
...
+
+    ```json
+    {
+        "op": "add",
+        "path": "groups",
+        "value": [
+                    {
+                    "value": "78144fd9-48e7-4fc9-95b5-cd3883f5ce4a"
+                }
+            ]
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:871]: formatting issue

File: en/docs/develop/scim2-patch-operations.md [hunk 25]
```diff
@@ -0,0 +1,928 @@
...
+    }
+    ```
+
+### Replace user groups
+
+Let's create patch request payloads to `replace` user groups assigned to a user role.
+
+-   **Example 1:** Replace a user group in a role (without using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:879]:
```suggestion
-   **Example 1:** Replace user groups in a role (without using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 26]
```diff
@@ -0,0 +1,928 @@
...
+            "groups": [
+                {
+                    "value": "78144fd9-48e7-4fc9-95b5-cd3883f5ce4a"
+                }
+            ]
+        }
+    }
+    ```
+
+-   **Example 2:** Replace a user group in a role (using the path param):
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:894]:
```suggestion
-   **Example 2:** Replace user groups in a role (using the path param):
```

File: en/docs/develop/scim2-patch-operations.md [hunk 27]
```diff
@@ -0,0 +1,928 @@
...
+
+    ```json
+    {
+        "op": "replace",
+        "path": "groups",
+        "value": [
+                    {
+                    "value": "78144fd9-48e7-4fc9-95b5-cd3883f5ce4a"
+                }
+            ]
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:904]: formatting issue

File: en/docs/develop/scim2-patch-operations.md [hunk 28]
```diff
@@ -0,0 +1,928 @@
...
+    ```
+
+-   **Example 2:** Specify the group (by group name) that needs to be removed.
+
+    ```json
+    {
+        "op": "remove",
+        "path": "groups[display eq PRIMARY/ABCD]"
+    }
+    ```
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-patch-operations.md:928]: Similar to Group rename, let's add Role renaming as well

File: en/docs/develop/scim2-batch-operations.md [hunk 29]
```diff
@@ -0,0 +1,1178 @@
...
+       </tr>
+       <tr>
+           <td><code>path</code></td>
+           <td>Required</td>
+           <td>
+               Add this path to specify the existing user group that should be deleted.</br>
+               <b>Possible value</b>:<code>/Groups/{group_id}</code>.
+           </td>
+       </tr>
+    </table>
```
Comment (AnuradhaSK) [inline en/docs/develop/scim2-batch-operations.md:1178]: Need to mention 
NOTE: Even though the multiple resources in the same resource type are managed per request in the above examples, it is allowed to do different operations (POST, PATCH, PUT, DELETE) on different resources (Users, Groups, …) in a single POST /Bulk operation.

Comment (AnuradhaSK) [inline en/docs/develop/scim2-batch-operations.md:1178]: Bulk role creation, update and remove also need to be included

File: en/docs/develop/scim2-batch-operations.md [hunk 30]
```diff
@@ -0,0 +1,1178 @@
...
+       <td><code>operations</code></td>
+       <td>Required</td>
+       <td>Specify the array of POST, PUT, PATCH, and/or DELETE operations that should be used to update the information. You can include any number of operations in one bulk request.</br></br>
+       See the instructions given below to define the required operations.</td>
+   </tr>
+</table>
+ 
+## Manage users in bulk
+ 
+You can use the **bulk** operations to add, remove, update, and delete users in bulk.
```
[the remark this answers is not in the document]
Comment (nilminiwso2) [reply en/docs/develop/scim2-batch-operations.md:39]: Should be `add, remove, update, and replace users in bulk.` Isn't it?

File: en/docs/develop/scim2-batch-operations.md (hunk 29 above)
Comment (nilminiwso2) [reply en/docs/develop/scim2-batch-operations.md:1205]: Added this note in two places:

```
The examples given below show individual resource operations (POST, PATCH, PUT, or DELETE) handled in a single request. However, note that a single request can execute a combination of operation types simultaneously.
```

Comment (nilminiwso2) [reply en/docs/develop/scim2-batch-operations.md:1205]: Created a separate issue for this one: https://github.com/wso2/product-is/issues/13110

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**